pytestCheckHook: support inclusion and exclusion of path globs, test items, keywords, and markers (#386513)
NixOS/nixpkgs: doc/languages-frameworks/python.section.md

Most packages use `pytest` or `unittest`, which is compatible with `pytest`,
so you will most likely use `pytestCheckHook`.

> To use `pytestCheckHook`, add it to `nativeCheckInputs`. Adding `pytest` explicitly is not required.

Reviewer: ```suggestion
To use `pytestCheckHook`, add it to `nativeCheckInputs`.
Adding `pytest` is not required, since it is included with `pytestCheckHook`.
```

```nix
{
  nativeCheckInputs = [
    pytestCheckHook
  ];
}
```

> `pytestCheckHook` recognizes the following attributes to include or exclude tests:

Reviewer: We could probably simplify this down to:
```suggestion
`pytestCheckHook` recognizes the following attributes:
```

> - `enabledTestPaths` and `disabledTestPaths`: to specify path globs (files or directories) or test items.
> - `enabledTests` and `disabledTests`: to specify keywords for class names or test method names.
> - `enabledTestMarks` and `disabledTestMarks`: to specify test marks.

Reviewer: Should we be using definition-list syntax here?

```suggestion
`enabledTestPaths` and `disabledTestPaths`

:   To specify path globs (files or directories) or test items.

`enabledTests` and `disabledTests`

:   To specify keywords for class names or test method names.

`enabledTestMarks` and `disabledTestMarks`

:   To specify test marks.
```

By default, `pytest` automatically discovers which tests to run.
If tests are explicitly enabled, only those tests will run.
A test, that is both enabled and disabled, will not run.

> The following example demonstrates the use of the "enabled" and "disabled" pytestCheckHook attributes:

Reviewer: ```suggestion
The following example demonstrates usage of various `pytestCheckHook` attributes:
```

```nix
{
  nativeCheckInputs = [
    pytestCheckHook
  ];

  # Allow running the following test paths and test objects.
  enabledTestPaths = [
    # Find tests under the tests directory.
    # The trailing slash is not necessary.
    "tests/"

    # Additionally run test_foo
    "other-tests/test_foo.py::Foo::test_foo"
  ];

  # Override the above-enabled test paths and test objects.
  disabledTestPaths = [
    # Tests under tests/integration requires additional data.
    "tests/integration"
  ];

  # Allow tests by keywords matching their class names or method names.
  enabledTests = [
    # pytest by default only runs test methods begin with "test_" or end with "_test".
    # This includes all functions whose name contains "test".
    "test"
  ];

  # Override the above-enabled tests by keywords matching their class names or method names.
  disabledTests = [
    # Tests touching networks.
    "upload"
    "download"
  ];
}
```

> The `<enabled/disabled>TestsPaths` expands Unix-style globs.

Reviewer: ```suggestion
The `<enabled/disabled>TestsPaths` attributes expand Unix-style globs.
```

> If the test paths to include/exclude contain characters like `*`, `?`, `[`, and `]`,

Reviewer: ```suggestion
If a test path contains characters like `*`, `?`, `[`, or `]`, you can
```

quote them with square brackets (`[*]`, `[?]`, `[[]`, and `[]]`) to match literally.

The `<enabled/disabled>Tests` and `<enabled/disabled>TestMarks` attribute pairs
form a logical expression `((included_element1) or (included_element2)) and not (excluded_element1) and not (excluded_element2)`
which will be passed to pytest's `-k` and `-m` flags respectively.
With `__structuredAttrs = true` enabled, they additionally support sub-expressions.

> For example,
> the following `disabledTests` specification excludes test items like
> `TestFoo::test_bar_functionality`:

Reviewer: ```suggestion
For example, you could disable test items like `TestFoo::test_bar_functionality`
by disabling tests that match both `"Foo"` **and** `"bar"`:
```

```nix
{
  __structuredAttrs = true;

  disabledTests = [
    "Foo and bar"
  ];
}
```

`pytestCheckHook` supports `pytestFlags` to append command-line arguments to `pytest`.
It requires `__structuredAttrs = true` to pass arguments containing spaces.

```nix
{
  # Additional pytest flags
  pytestFlags = [
    # Disable benchmarks and run benchmarking tests only once.
    "--benchmark-disable"
  ];
}
```

> The main benefits of constructing `pytest` commands from `pytestCheckHook`-supported attributes/variables

Reviewer: ```suggestion
The main benefits of using `pytestCheckHook` to construct `pytest` commands
```

Most packages use `pytest` or `unittest`, which is compatible with `pytest`,
so you will most likely use `pytestCheckHook`.

To use `pytstCheckHook`, add it to `nativeCheckInputs`. Adding `pytest` explicitly is not required.

```nix
{
  nativeCheckInputs = [
    pytestCheckHook
  ];
}
```

`pytestCheckHook` recognizes the following attributes for whitelisting and/or blacklisting tests:

- `enabledTestPaths` and `disabledTestPaths`: to specify path globs (files or directories) or test items.
- `enabledTests` and `disabledTests`: to specify keywords for class names or test method names.
- `enabledTestMarks` and `disabledTestMarks`: to specify test marks.

"enabled" attributes define a whitelist of tests to run, while "disabled" attributes define a blacklist of tests that will not be run.
If the whitelist is empty, it is ignored and only the blacklist is considered.

> If a test matches both the whitelist and blacklist, it is not run.

Author: That sounds a lot clearer!

Nevertheless, "all tests" here might refer to "all the tests included during pytest's test discovery." If not further configured, it requires matching pytest's default test naming conventions:

```toml
[tool.pytest.ini_options]
python_files = [
    "test_*.py",
    "*_test.py",
]
python_classes = [
    "Test*",
]
pthon_functions = [
    "test*",
]
```

where `python_functions` governs module-level functions or class methods.

The `enable*` attributes are able to add tests outside such naming conventions.

The following example demonstrates the use of the "enabled" and "disabled" pytestCheckHook attributes:

```nix
{
  nativeCheckInputs = [
    pytestCheckHook
  ];

  # Whitelist the test paths and test objects.
  enabledTestPaths = [
    # Find tests under the tests directory.
    # The trailing slash is not necessary.
    "tests/"

    # Additionally run test_foo
    "other-tests/test_foo.py::Foo::test_foo"
  ];

  # Blacklist the test paths and test objects
  disabledTestPaths = [
    # Tests under tests/integration requires additional data.
    "tests/integration"
  ];

  # Whitelist tests by keywords matching their class names or method names.
  enabledTests = [
    # pytest by default only runs test methods begin with "tests_"
    "test"
  ];

  # Blacklist tests by keywords matching their class names or method names.
  disabledTests = [
    # Tests touching networks.
    "upload"
    "download"
  ];

>   # Blacklist tests by test marks.

Author: I haven'found a real-world example, but specifying `enabledTestMarks` is likely to exclude most of the tests otherwise included.

Still, it is occationally practical to include only a small part of the tests. I just cannot come up with a value that goes well with other attributes values specifed here.

>   # Blacklist tests by test marks.

Author: > I guess everybody has understood the concept by here, but it's inconsistent. We could probably skip some others, too.

How about we remove `disabledTestMarks` from  the example? The implementation logic is similar to `enabledTests` and `disabledTests`, and maybe `enabledTests` and `disabledTests` are enough to show how to use them.

>   # Blacklist tests by test marks.

Author: > > How about we remove `disabledTestMarks` from the example?
>
> Fine by me!

Done.

Most packages use `pytest` or `unittest`, which is compatible with `pytest`,
so you will most likely use `pytestCheckHook`.

To use `pytstCheckHook`, add it into the `nativeBuildInputs`. (You don't have to add `pytest` explicitly for `pytestCheckHook` to work.)

```nix
{
  nativeCheckInputs = [
    pytestCheckHook
  ];
}
```

`pytestCheckHook` recognizes the following attributes/variables to include/whitelist tests and to exclude/blacklist tests:

> - `enabledTestPaths` and `disabledTestPath`: to specify path globs (files or directories) or test items.

Reviewer 2: ```suggestion
- `enabledTestPaths` and `disabledTestPaths`: to specify path globs (files or directories) or test items.
```

Most packages use `pytest` or `unittest`, which is compatible with `pytest`,
so you will most likely use `pytestCheckHook`.

> To use `pytstCheckHook`, add it into the `nativeBuildInputs`. (You don't have to add `pytest` explicitly for `pytestCheckHook` to work.)

Reviewer 3: ```suggestion
To use `pytstCheckHook`, add it to `nativeBuildInputs`. Adding `pytest` explicitly is not required.
```

Author: > Should be `nativeCheckInputs` instead of `nativeBuildInputs`, right?

You're right.

To use `pytstCheckHook`, add it into the `nativeBuildInputs`. (You don't have to add `pytest` explicitly for `pytestCheckHook` to work.)

```nix
{
  nativeCheckInputs = [
    pytestCheckHook
  ];

  # requires additional data
  pytestFlags = [
    "tests/"
    "--ignore=tests/integration"
  ];

  disabledTests = [
    # touches network
    "download"
    "update"
  ];

  disabledTestPaths = [
    "tests/test_failing.py"

>   ];

Author: The original example emphasize the transition from using `pytest` command inside `checkPhase` (`installCheckPhase`) to `pytestCheckHook` and is coupled to the historical implementation of the hook. Let me add a new one exclusively for `pytestCheckHook`.

>   ];

Author: Addressed.

> `pytestCheckHook` recognizes the following attributes/variables to include/whitelist tests and to exclude/blacklist tests:

Reviewer: Simplified wording:
```suggestion
`pytestCheckHook` recognizes the following attributes for whitelisting and/or blacklisting tests:
```

> 

Reviewer: ```suggestion

  "enabled" attributes define a whitelist of tests to run, while "disabled" attributes define a blacklist of tests that will not be run.
  If the whitelist is empty, it is ignored and only the blacklist is considered.
  If a test matches both the whitelist and blacklist, it is not run.

```

  __structuredAttrs = true;

  disabledTests = [

>     "Foo and bar"

Reviewer: Again, why remove real world examples?

Author: I added it back.
